Index: mercurial-0.7/setup.py
===================================================================
--- mercurial-0.7.orig/setup.py	2005-09-23 11:33:26.000000000 +0200
+++ mercurial-0.7/setup.py	2005-09-23 11:35:02.000000000 +0200
@@ -29,7 +29,7 @@
           url='http://selenic.com/mercurial',
           description='scalable distributed SCM',
           license='GNU GPL',
-          packages=['mercurial'],
+          packages=['mercurial', 'hgext'],
           ext_modules=[Extension('mercurial.mpatch', ['mercurial/mpatch.c']),
                        Extension('mercurial.bdiff', ['mercurial/bdiff.c'])],
           data_files=[('mercurial/templates',
