Index: mercurial-0.7/contrib/hgit
===================================================================
--- mercurial-0.7.orig/contrib/hgit	2005-09-23 10:34:30.000000000 +0200
+++ mercurial-0.7/contrib/hgit	2005-09-23 10:42:15.000000000 +0200
@@ -7,7 +7,7 @@
 # This software may be used and distributed according to the terms
 # of the GNU General Public License, incorporated herein by reference.
 
-import time, sys, signal
+import time, sys, signal, os
 from mercurial import hg, mdiff, fancyopts, commands, ui, util
 
 def dodiff(fp, ui, repo, node1, node2, files=None, match=util.always,
@@ -86,7 +86,7 @@
 
         for f in c:
             # TODO get file permissions
-            print ":100664 100664 %s %s M\t%s\t%s" % (hg.hex(mmap[f]), 
+            print ":100664 100664 %s %s M\t%s\t%s" % (hg.hex(mmap[f]),
                                                       hg.hex(mmap2[f]), f, f)
         for f in a:
             print ":000000 100664 %s %s N\t%s\t%s" % (empty, hg.hex(mmap2[f]), f, f)
@@ -137,7 +137,7 @@
         committer = lines[-1].split(': ')[1].rstrip()
     else:
         committer = "%s %s %s" % (changes[1], date, date_ar[1])
-        
+
     print "author %s %s %s" % (changes[1], date, date_ar[1])
     print "committer %s" % (committer)
     print ""
@@ -220,7 +220,7 @@
                     yield (i + x, full != None and l[x] or None)
             if i == 0:
                 break
-            
+
     # calculate and return the reachability bitmask for sha
     def is_reachable(ar, reachable, sha):
         if len(ar) == 0:
@@ -313,22 +313,28 @@
     copy = [x for x in revs]
     revtree(copy, repo, full, opts['max_count'], opts['parents'])
 
+def view(ui, repo, *etc):
+    "start interactive history viewer"
+    os.chdir(repo.root)
+    os.system("hgk " + " ".join(etc))
+
 cmdtable = {
-    "git-diff-tree": (difftree, [('p', 'patch', None, 'generate patch'),
+    "view": (view, [], 'hg view'),
+    "debug-diff-tree": (difftree, [('p', 'patch', None, 'generate patch'),
                             ('r', 'recursive', None, 'recursive'),
                             ('P', 'pretty', None, 'pretty'),
                             ('s', 'stdin', None, 'stdin'),
                             ('C', 'copy', None, 'detect copies'),
                             ('S', 'search', "", 'search')],
                             "hg git-diff-tree [options] node1 node2"),
-    "git-cat-file": (catfile, [('s', 'stdin', None, 'stdin')],
-                 "hg cat-file [options] type file"),
-    "git-merge-base": (base, [], "hg git-merge-base node node"),
-    "git-rev-list": (revlist, [('H', 'header', None, 'header'),
+    "debug-cat-file": (catfile, [('s', 'stdin', None, 'stdin')],
+                 "hg debug-cat-file [options] type file"),
+    "debug-merge-base": (base, [], "hg debug-merge-base node node"),
+    "debug-rev-list": (revlist, [('H', 'header', None, 'header'),
                            ('t', 'topo-order', None, 'topo-order'),
                            ('p', 'parents', None, 'parents'),
                            ('n', 'max-count', 0, 'max-count')],
-                 "hg git-rev-list [options] revs"),
+                 "hg debug-rev-list [options] revs"),
 }
 
 def reposetup(ui, repo):
Index: mercurial-0.7/contrib/hgk
===================================================================
--- mercurial-0.7.orig/contrib/hgk	2005-09-16 23:02:33.000000000 +0200
+++ mercurial-0.7/contrib/hgk	2005-09-23 10:41:53.000000000 +0200
@@ -34,7 +34,7 @@
     set ncmupdate 1
     if [catch {
 	set parse_args [concat --default HEAD $rargs]
-	set parsed_args [split [eval exec hg git-rev-parse $parse_args] "\n"]
+	set parsed_args [split [eval exec hg debug-rev-parse $parse_args] "\n"]
     }] {
 	# if git-rev-parse failed for some reason...
 	if {$rargs == {}} {
@@ -43,9 +43,9 @@
 	set parsed_args $rargs
     }
     if [catch {
-	set commfd [open "|hg git-rev-list --header --topo-order --parents $parsed_args" r]
+	set commfd [open "|hg debug-rev-list --header --topo-order --parents $parsed_args" r]
     } err] {
-	puts stderr "Error executing hg git-rev-list: $err"
+	puts stderr "Error executing hg debug-rev-list: $err"
 	exit 1
     }
     set leftover {}
@@ -113,7 +113,7 @@
 	    if {[string length $shortcmit] > 80} {
 		set shortcmit "[string range $shortcmit 0 80]..."
 	    }
-	    error_popup "Can't parse hg git-rev-list output: {$shortcmit}"
+	    error_popup "Can't parse hg debug-rev-list output: {$shortcmit}"
 	    exit 1
 	}
 	set id [lindex $ids 0]
@@ -164,7 +164,7 @@
 }
 
 proc readcommit {id} {
-    if [catch {set contents [exec hg git-cat-file commit $id]}] return
+    if [catch {set contents [exec hg debug-cat-file commit $id]}] return
     parsecommit $id $contents 0 {}
 }
 
@@ -1708,7 +1708,7 @@
     }
 
     if {[catch {
-	set f [open [list | hg git-diff-tree --stdin -s -r -S$findstring \
+	set f [open [list | hg debug-diff-tree --stdin -s -r -S$findstring \
 			 << $inputids] r]
     } err]} {
 	error_popup "Error starting search process: $err"
@@ -1811,7 +1811,7 @@
     # start off a git-diff-tree process if needed
     if {$diffsneeded ne {}} {
 	if {[catch {
-	    set df [open [list | hg git-diff-tree -r --stdin << $diffsneeded] r]
+	    set df [open [list | hg debug-diff-tree -r --stdin << $diffsneeded] r]
 	} err ]} {
 	    error_popup "Error starting search process: $err"
 	    return
@@ -1843,7 +1843,7 @@
 	    if {[catch {close $df} err]} {
 		stopfindproc
 		bell
-		error_popup "Error in hg git-diff-tree: $err"
+		error_popup "Error in hg debug-diff-tree: $err"
 	    } elseif {[info exists findids]} {
 		set ids $findids
 		stopfindproc
@@ -2113,7 +2113,7 @@
 	}
 	$ctext insert end "\n"
     }
- 
+
     set comment {}
     if {[info exists parents($id)]} {
 	foreach p $parents($id) {
@@ -2232,7 +2232,7 @@
 	    set gca $id
 	} else {
 	    if {[catch {
-		set gca [exec hg git-merge-base $gca $id]
+		set gca [exec hg debug-merge-base $gca $id]
 	    } err]} {
 		return {}
 	    }
@@ -2320,7 +2320,7 @@
     catch {unset grouphunks}
     set groupfilelast -1
     foreach p $parents($diffmergeid) {
-	set cmd [list | hg git-diff-tree -p $p $diffmergeid]
+	set cmd [list | hg debug-diff-tree -p $p $diffmergeid]
 	set cmd [concat $cmd $mergefilelist($diffmergeid)]
 	if {[catch {set f [open $cmd r]} err]} {
 	    error_popup "Error getting diffs: $err"
@@ -2764,7 +2764,7 @@
     set treediff {}
     set id [lindex $ids 0]
     set p [lindex $ids 1]
-    if [catch {set gdtf [open "|hg git-diff-tree -r $p $id" r]}] return
+    if [catch {set gdtf [open "|hg debug-diff-tree -r $p $id" r]}] return
     fconfigure $gdtf -blocking 0
     fileevent $gdtf readable [list gettreediffline $gdtf $ids]
 }
@@ -2800,7 +2800,7 @@
     set id [lindex $ids 0]
     set p [lindex $ids 1]
     set env(GIT_DIFF_OPTS) $diffopts
-    set cmd [list | hg git-diff-tree -r -p -C $p $id]
+    set cmd [list | hg debug-diff-tree -r -p -C $p $id]
     if {[catch {set bdf [open $cmd r]} err]} {
 	puts "error getting diffs: $err"
 	return
@@ -3387,7 +3387,7 @@
     set oldid [$patchtop.fromsha1 get]
     set newid [$patchtop.tosha1 get]
     set fname [$patchtop.fname get]
-    if {[catch {exec hg git-diff-tree -p $oldid $newid >$fname &} err]} {
+    if {[catch {exec hg debug-diff-tree -p $oldid $newid >$fname &} err]} {
 	error_popup "Error creating patch: $err"
     }
     catch {destroy $patchtop}
@@ -3608,7 +3608,7 @@
 set datemode 0
 set boldnames 0
 set diffopts "-U 5 -p"
-set wrcomcmd "hg git-diff-tree --stdin -p --pretty"
+set wrcomcmd "hg debug-diff-tree --stdin -p --pretty"
 
 set mainfont {Helvetica 9}
 set textfont {Courier 9}
