Index: mercurial-0.9/mercurial/commands.py
===================================================================
--- mercurial-0.9.orig/mercurial/commands.py	2006-05-12 19:30:18.000000000 +0200
+++ mercurial-0.9/mercurial/commands.py	2006-05-12 19:30:29.000000000 +0200
@@ -2280,6 +2280,7 @@
                            'you must specify the revision to revert to'))
     else:
         node = parent
+    pmf = None
     mf = repo.manifest.read(repo.changelog.read(node)[0])
 
     wlock = repo.wlock()
@@ -2362,7 +2363,12 @@
                 if exact: ui.warn(_('no changes needed to %s\n' % rel))
                 continue
             if not in_mf:
-                handle(remove, False)
+                if pmf is None:
+                    # only need parent manifest in this unlikely case,
+                    # so do not read by default
+                    pmf = repo.manifest.read(repo.changelog.read(parent)[0])
+                if abs in pmf:
+                    handle(remove, False)
         update[abs] = True
 
     repo.dirstate.forget(forget[0])
