--- mercurial-0.7/contrib/bash_completion	2005-09-16 23:02:33.000000000 +0200
+++ mercurial-0.7/contrib/bash_completion	2005-12-15 13:13:15.000000000 +0100
@@ -20,6 +20,14 @@
     COMPREPLY=(${COMPREPLY[@]:-} $( compgen -W "$paths" -- "$cur" ))
 }
 
+_hg_repos()
+{
+    local i
+    for i in $( compgen -d -- "$cur" ); do
+        test ! -d "$i"/.hg || COMPREPLY=(${COMPREPLY[@]:-} "$i")
+    done
+}
+
 _hg_status()
 {
     local files="$( hg status -$1 | cut -b 3- )"
@@ -83,11 +91,11 @@
     # global options
     case "$prev" in
 	-R|--repository)
-	    COMPREPLY=(${COMPREPLY[@]:-} $( compgen -d -- "$cur" ))
+	    _hg_repos
 	    return
 	;;
 	--cwd)
-	    COMPREPLY=(${COMPREPLY[@]:-} $( compgen -d -- "$cur" ))
+	    # Stick with default bash completion
 	    return
 	;;
     esac
@@ -114,7 +122,7 @@
 	;;
 	pull|push|outgoing|incoming)
 	    _hg_paths
-	    COMPREPLY=(${COMPREPLY[@]:-} $( compgen -d -- "$cur" ))
+	    _hg_repos
 	;;
 	paths)
 	    _hg_paths
@@ -142,7 +150,7 @@
 	    if [ $count = 1 ]; then
 		_hg_paths
 	    fi
-	    COMPREPLY=(${COMPREPLY[@]:-} $( compgen -d -- "$cur" ))
+            _hg_repos
 	;;
 	debugindex|debugindexdot)
 	    COMPREPLY=(${COMPREPLY[@]:-} $( compgen -f -X "!*.i" -- "$cur" ))
