Index: mercurial-0.7/mercurial/commands.py
===================================================================
--- mercurial-0.7.orig/mercurial/commands.py	2005-11-02 06:17:35.000000000 +0100
+++ mercurial-0.7/mercurial/commands.py	2005-11-22 12:59:44.000000000 +0100
@@ -1946,7 +1946,7 @@
           ('', 'style', "", 'template style'),
           ('6', 'ipv6', None, 'use IPv6 in addition to IPv4')],
          "hg serve [OPTION]..."),
-    "^status":
+    "^status|st":
         (status,
          [('m', 'modified', None, 'show only modified files'),
           ('a', 'added', None, 'show only added files'),
