Index: mercurial-0.7/mercurial/commands.py
===================================================================
--- mercurial-0.7.orig/mercurial/commands.py	2005-12-29 12:50:19.000000000 +0100
+++ mercurial-0.7/mercurial/commands.py	2006-01-03 20:44:05.000000000 +0100
@@ -629,9 +629,10 @@
     if opts['remotecmd']:
         ui.setconfig("ui", "remotecmd", opts['remotecmd'])
 
-    d = Dircleanup(dest)
     if not os.path.exists(source):
         source = ui.expandpath(source)
+
+    d = Dircleanup(dest)
     abspath = source
     other = hg.repository(ui, source)
 
