Index: mercurial-git/hgext/convert/__init__.py
===================================================================
--- mercurial-git.orig/hgext/convert/__init__.py	2008-07-08 20:55:52.000000000 +0200
+++ mercurial-git/hgext/convert/__init__.py	2008-07-08 20:56:16.000000000 +0200
@@ -16,7 +16,7 @@
     Accepted source formats:
     - Mercurial
     - CVS
-    - Darcs
+    - Darcs (legacy Darcs 1 format only)
     - git
     - Subversion
     - Monotone
