Index: mercurial-0.9.1/contrib/purge/purge.py
===================================================================
--- mercurial-0.9.1.orig/contrib/purge/purge.py	2006-07-25 19:06:27.000000000 +0200
+++ mercurial-0.9.1/contrib/purge/purge.py	2006-07-25 19:06:32.000000000 +0200
@@ -1,5 +1,3 @@
-#!/usr/bin/env python
-#
 # Copyright (C) 2006 - Marco Barisione <marco@barisione.org>
 #
 # This is a small extension for Mercurial (http://www.selenic.com/mercurial)
