Index: mercurial-git/hgext/convert/darcs.py
===================================================================
--- mercurial-git.orig/hgext/convert/darcs.py	2008-04-06 22:05:27.000000000 +0200
+++ mercurial-git/hgext/convert/darcs.py	2008-04-06 22:11:01.000000000 +0200
@@ -30,7 +30,7 @@
         checktool('darcs')
 
         if ElementTree is None:
-            raise util.Abort(_("Python ElementTree module is not available"))
+            raise util.Abort(_("Python ElementTree module is not available. Try installing the python-elementtree Debian package to convert from darcs"))
 
         if not os.path.exists(os.path.join(path, '_darcs', 'inventory')):
             raise NoRepo("%s does not look like a darcs repo" % path)
