Index: mercurial-0.8.1/contrib/hgk.py
===================================================================
--- mercurial-0.8.1.orig/contrib/hgk.py	2006-04-08 04:23:11.000000000 +0200
+++ mercurial-0.8.1/contrib/hgk.py	2006-04-10 20:26:35.000000000 +0200
@@ -314,7 +314,7 @@
 def view(ui, repo, *etc):
     "start interactive history viewer"
     os.chdir(repo.root)
-    os.system(ui.config("hgk", "path", "hgk") + " " + " ".join(etc))
+    os.system(ui.config("hgk", "path", "/usr/share/mercurial/hgk") + " " + " ".join(etc))
 
 cmdtable = {
     "view": (view, [], 'hg view'),
