diff -r 434139080ed4 doc/hgrc.5.txt
--- a/doc/hgrc.5.txt	Mon Jan 28 22:19:12 2008 -0500
+++ b/doc/hgrc.5.txt	Tue Feb 26 08:58:59 2008 -0600
@@ -512,7 +512,7 @@ ui::
   debug;;
     Print debugging information.  True or False.  Default is False.
   editor;;
-    The editor to use during a commit.  Default is $EDITOR or "vi".
+    The editor to use during a commit.  Default is $EDITOR or "sensible-editor".
   fallbackencoding;;
     Encoding to try if it's not possible to decode the changelog using
     UTF-8.  Default is ISO-8859-1.
diff -r 434139080ed4 doc/ja/hg.1.ja.txt
--- a/doc/ja/hg.1.ja.txt	Mon Jan 28 22:19:12 2008 -0500
+++ b/doc/ja/hg.1.ja.txt	Tue Feb 26 08:58:59 2008 -0600
@@ -813,7 +813,7 @@ EDITOR::
 EDITOR::
     これは hgmerge スクリプト中で使われるエディタの名前です。もし
     HGEDITOR が設定されていなければ、コミット時のメッセージに使わ
-    れます。デフォルトは 'vi' です。
+    れます。デフォルトは 'sensible-editor' です。
 
 PYTHONPATH::
     これはインポートされるモジュールを見つけるために Python によっ
diff -r 434139080ed4 doc/ja/hgrc.5.ja.txt
--- a/doc/ja/hgrc.5.ja.txt	Mon Jan 28 22:19:12 2008 -0500
+++ b/doc/ja/hgrc.5.ja.txt	Tue Feb 26 08:58:59 2008 -0600
@@ -119,7 +119,7 @@ ui::
     では False です。
   editor;;
     コミット中に使用するエディタです。デフォルトは $EDITOR か
-    "vi" です。
+    "sensible-editor" です。
   interactive;;
     ユーザに対してプロンプトを出すようにします。True か False を取
     ります。デフォルトでは True です。
diff -r 434139080ed4 hgeditor
--- a/hgeditor	Mon Jan 28 22:19:12 2008 -0500
+++ b/hgeditor	Tue Feb 26 08:58:59 2008 -0600
@@ -7,7 +7,7 @@
 # only for Mercurial, modify this:
 case "${EDITOR}" in
     "")
-        EDITOR="vi"
+        EDITOR="sensible-editor"
         ;;
     emacs)
         EDITOR="$EDITOR -nw"
diff -r 434139080ed4 mercurial/commands.py
--- a/mercurial/commands.py	Mon Jan 28 22:19:12 2008 -0500
+++ b/mercurial/commands.py	Tue Feb 26 08:58:59 2008 -0600
@@ -835,8 +835,8 @@ def debuginstall(ui):
     editor = ui.geteditor()
     cmdpath = util.find_exe(editor) or util.find_exe(editor.split()[0])
     if not cmdpath:
-        if editor == 'vi':
-            ui.write(_(" No commit editor set and can't find vi in PATH\n"))
+        if editor == 'sensible-editor':
+            ui.write(_(" No commit editor set and can't find sensible-editor in PATH\n"))
             ui.write(_(" (specify a commit editor in your .hgrc file)\n"))
         else:
             ui.write(_(" Can't find editor '%s' in PATH\n") % editor)
diff -r 434139080ed4 mercurial/ui.py
--- a/mercurial/ui.py	Mon Jan 28 22:19:12 2008 -0500
+++ b/mercurial/ui.py	Tue Feb 26 08:58:59 2008 -0600
@@ -476,5 +476,5 @@ class ui(object):
         return (os.environ.get("HGEDITOR") or
                 self.config("ui", "editor") or
                 os.environ.get("VISUAL") or
-                os.environ.get("EDITOR", "vi"))
+                os.environ.get("EDITOR", "sensible-editor"))
 
