Set default hgk path for hgk outside bin.

Index: mercurial-4.4.1/hgext/hgk.py
===================================================================
--- mercurial-4.4.1.orig/hgext/hgk.py	2017-11-13 07:03:35.564787794 +0200
+++ mercurial-4.4.1/hgext/hgk.py	2017-11-13 07:04:16.637499141 +0200
@@ -65,7 +65,7 @@
 configitem = registrar.configitem(configtable)
 
 configitem('hgk', 'path',
-    default='hgk',
+    default='/usr/share/mercurial/hgk',
 )
 
 @command('debug-diff-tree',
