fix handling of daylight saving time
Index: mercurial-0.7/mercurial/changelog.py
===================================================================
--- mercurial-0.7.orig/mercurial/changelog.py	2005-09-16 23:02:33.000000000 +0200
+++ mercurial-0.7/mercurial/changelog.py	2005-11-02 06:15:40.000000000 +0100
@@ -44,9 +44,12 @@
             if abs(offset) >= 43200:
                 raise ValueError('impossible time zone offset: %d' % offset)
         else:
-            if time.daylight: offset = time.altzone
-            else: offset = time.timezone
-            date = "%d %d" % (time.time(), offset)
+            lt = time.localtime()
+            if lt[8] == 1 and time.daylight:
+                tz = time.altzone
+            else:
+                tz = time.timezone
+            date = "%d %d" % (time.mktime(lt), tz)
         list.sort()
         l = [hex(manifest), user, date] + list + ["", desc]
         text = "\n".join(l)
