Author: Matt Mackall <mpm@selenic.com>
Description: tests: re-silence test-bad-pull log messages
Origin: upstream, http://selenic.com/hg/rev/91eec29dd7de

diff --git a/tests/test-bad-pull.t b/tests/test-bad-pull.t
--- a/tests/test-bad-pull.t
+++ b/tests/test-bad-pull.t
@@ -25,6 +25,3 @@
   abort: HTTP Error 404: * (glob)
   [255]
   $ wait $P
-  $ cat log
-  localhost - - [*] code 404, message File not found (glob)
-  localhost - - [*] "GET /foo?cmd=capabilities HTTP/1.1" 404 - (glob)
