Index: mercurial-0.8.1/hgext/hbisect.py
===================================================================
--- mercurial-0.8.1.orig/hgext/hbisect.py	2006-04-08 04:23:11.000000000 +0200
+++ mercurial-0.8.1/hgext/hbisect.py	2006-04-10 21:24:46.000000000 +0200
@@ -246,7 +246,7 @@
     return 0
 
 def bisect_run(ui, repo, cmd=None, *args):
-    """bisect extension: dichotomic search in the DAG of changesets
+    """dichotomic search in the DAG of changesets
 for subcommands see "hg bisect help\"
     """
     def help_(cmd=None, *args):
