Index: mercurial-0.7+20060110+0d36e3d7e2ea/contrib/hgk.py
===================================================================
--- mercurial-0.7+20060110+0d36e3d7e2ea.orig/contrib/hgk.py	2005-09-29 10:51:11.000000000 +0200
+++ mercurial-0.7+20060110+0d36e3d7e2ea/contrib/hgk.py	2006-01-10 21:42:05.000000000 +0100
@@ -1,5 +1,3 @@
-#!/usr/bin/env python
-#
 # Minimal support for git commands on an hg repository
 #
 # Copyright 2005 Chris Mason <mason@suse.com>
