Index: mercurial-0.8.1/contrib/hgk
===================================================================
--- mercurial-0.8.1.orig/contrib/hgk	2006-04-11 13:28:43.000000000 +0200
+++ mercurial-0.8.1/contrib/hgk	2006-04-12 20:33:14.000000000 +0200
@@ -3448,7 +3448,7 @@
 	return
     }
     if {[catch {
-	set out [exec hg tag $tag $id]
+	set out [exec hg tag -r $id $tag]
     } err]} {
 	error_popup "Error creating tag: $err"
 	return
